pubsub-mapped-api/src/test/java/com/google/pubsub/clients/consumer/KafkaConsumerTest.java

```diff
@@ -0,0 +1,333 @@
+
+    consumer.subscribe(Arrays.asList("topic", "topic1", "topic"));
+
+    Set<String> subscribed = consumer.subscription();
+    assertTrue(topics.equals(subscribed));
+  }
+
+  @Test
+  public void testUnsubscribeSubscriptionDeleted() {
+    grpcServerRule.getServiceRegistry().addService(new SubscriberImpl());
+
+    Set<String> topics = new HashSet<>(Arrays.asList("topic", "topic1"));
+
+    consumer.subscribe(Arrays.asList("topic", "topic1", "topic"));
+
+    Set<String> subscribed = consumer.subscription();
+    assertTrue(topics.equals(subscribed));
+  }
+
+  @Test
+  public void testPatternSubscription() {
+    grpcServerRule.getServiceRegistry().addService(new SubscriberImpl());
+    grpcServerRule.getServiceRegistry().addService(new PublisherImpl());
+
+    Pattern pattern = Pattern.compile("[a-z]*\\d{3}cat");
+
+    consumer.subscribe(pattern, null);
+    Set<String> subscribed = consumer.subscription();
+    Set<String> expectedTopics = new HashSet<>(Arrays.asList("thisis123cat",
+        "funnycats000cat"));
+    assertEquals(expectedTopics, subscribed);
+  }
+
+  @Test
+  public void emptyPatternFails() {
+    try {
+      Pattern pattern = null;
+      consumer.subscribe(pattern, null);
+      fail();
+    } catch (IllegalArgumentException e) {
+      assertEquals("Topic pattern to subscribe to cannot be null", e.getMessage());
+    }
+  }
+
+  @Test
+  public void nullTopicListFails() {
+    try {
+      List<String> topics = null;
+      consumer.subscribe(topics);
+      fail();
+    } catch (IllegalArgumentException e) {
+      assertEquals("Topic collection to subscribe to cannot be null", e.getMessage());
+    }
+  }
+
+  @Test
+  public void emptyTopicFails() {
+    try {
+      List<String> topics = new ArrayList<>(Collections.singletonList("    "));
+      consumer.subscribe(topics);
+      fail();
+    } catch (IllegalArgumentException e) {
+      assertEquals("Topic collection to subscribe to cannot contain null or empty topic",
+          e.getMessage());
+    }
+  }
+
+  @Test
+  public void noSubscriptionsPollFails() {
+    try {
+      consumer.poll(100);
+      fail();
+    } catch (IllegalStateException e) {
+      assertEquals("Consumer is not subscribed to any topics", e.getMessage());
+    }
+  }
+
+  @Test
+  public void negativePollTimeoutFails() {
+    try {
+      consumer.poll(-200);
+      fail();
+    } catch (IllegalArgumentException e) {
+      assertEquals("Timeout must not be negative", e.getMessage());
+    }
+  }
+
+
+  @Test
+  public void testDeserializeProperly() throws ExecutionException, InterruptedException {
+    grpcServerRule.getServiceRegistry().addService(new SubscriberImpl());
+
+    String topic = "topic";
+    Integer key = 125;
+    String value = "value";
+    consumer.subscribe(Collections.singletonList(topic));
+
+    ConsumerRecords<Integer, String> consumerRecord = consumer.poll(100);
+
+    assertEquals(consumerRecord.count(), 1);
```

Comment:
it's hard to see in this code what is being tested.  Mixing `assert*` statements rather than making them all at the bottom makes it seem like this method is testing too much.